File: Javascript/Algorithms/1046.js
let lastStoneWeight = stones => {
    let pq = new MaxPriorityQueue({priority: x => x});
    stones.forEach(x => pq.enqueue(x));
    
    while (1){
        let [e1, e2] = [pq.dequeue(), pq.dequeue()];
        if (!e1){
            return 0;
        }
        if (!e2){
            return e1.element;
        }
        let diff = e1.element - e2.element;
        if (diff){
            pq.enqueue(diff);
        }
    }
};

File: C/Algorithms/1046.c
#include <stdlib.h>

int cmpi(const void *x, const void *y) {
  int a = *(const int *)x;
  int b = *(const int *)y;

  return (a > b) - (a < b);
}

int lastStoneWeight(int *stones, int stonesSize) {
  while (1) {
    qsort(stones, stonesSize, sizeof(int), cmpi);
    if (stonesSize >= 2) {
      int e1 = stones[stonesSize - 1];
      int diff = e1 - stones[stonesSize - 2];
      if (diff) {
        stones[stonesSize - 2] = diff;
        stonesSize -= 1;
      } else
        stonesSize -= 2;
    } else {
      if (stonesSize == 0)
        return 0;
      else if (stonesSize == 1) {
        return stones[0];
      }
    }
  }
  // unreachable.
  return 0;
}
